### ISSUE TITLE
Destroying element parent to context manager-handled dialog crashes wx on Mac OS

### ISSUE THREAD

**Issue by DarkFenX** [wxWidgets/Phoenix#1343]
**Operating system**: Mac OS 10.14.6
**wxPython version & source**: 4.0.6 from pip
**Python version & source**: Python 3.7.4, I believe bundled with the OS

**Description of the problem**:
1. Create some Window element within a Frame
2. Make this Window to spawn Dialog via context manager
3. As effect of the dialog, Destroy the Window (within context manager or shortly after it)

```python
import wx


class MainFrame(wx.Frame):

    def __init__(self):
        super().__init__(None, wx.ID_ANY, 'title')
        sizer = wx.BoxSizer(wx.HORIZONTAL)
        item = Item(self)
        sizer.Add(item, 1, wx.EXPAND)
        self.SetSizer(sizer)
        self.Layout()
        self.Show()


class Item(wx.Window):

    def __init__(self, parent):
        wx.Window.__init__(self, parent)
        sizer = wx.BoxSizer(wx.HORIZONTAL)
        deleteBtn = wx.Button(self, label='Destroy me')
        sizer.Add(deleteBtn, 1, wx.EXPAND)
        self.SetSizer(sizer)
        self.Layout()
        deleteBtn.Bind(wx.EVT_BUTTON, self.OnDelete)


    def OnDelete(self, event):
        with wx.MessageDialog(self, 'Really?', 'Confirm', wx.YES | wx.NO | wx.ICON_QUESTION ) as dlg:
            if dlg.ShowModal() == wx.ID_YES:
                self.delete()

    def delete(self):
        self.Destroy()


if __name__ == '__main__':
    app = wx.App(False)
    mf = MainFrame()
    app.MainLoop()
```

On Mac, when you click "Destroy me" button and confirm, app crashes. On windows/linux, it does not.

If i move Window destruction out of context manager:
```
    def OnDelete(self, event):
        destroy = False
        with wx.MessageDialog(self, 'Really?', 'Confirm', wx.YES | wx.NO | wx.ICON_QUESTION ) as dlg:
            if dlg.ShowModal() == wx.ID_YES:
                destroy = True
        if destroy: self.delete()
```
It still crashes.

Is it a bug, or maybe I am doing it wrong?

**Comment by RobinD42** [comment]
Basically it is happening because there is a double `Destroy` of the dialog. Once when the context manager exits, and once when the `Item` destroys its children. It still happens when you move the `self.delete()` call outside of the context manager because the `Destroy` of top-level windows is deferred until the the message queue empties to make sure there are no pending events for it. So either way you end up with two items on the pending delete queue for 1 dialog.

I'll give it some more thought but I don't think there is anything I can do in the wxPython layers that would cover all cases of the problem. There is a simple workaround however. Don't make the dialog be a child of the `Item`. You can use the parent frame as the dialog's parent, or just pass `None`.

**Comment by DarkFenX** [comment]
I changed my code to not using context manager, posting event which destroys parent if user clicks OK and destroying the dialog explicitly afterwards. It works, but is it acceptable solution from wxpython's coding guidelines point of view?

**Comment by RobinD42** [comment]
If I understand correctly then it's probably okay, although there is nothing preventing a change in the order of processing the posted event and running though the pending delete queue sometime in the future, so there could possibly still be some fragility there.

Is there a reason you need the dialog to be a child of the Item window? Would using `self.GetTopLevelParent()` or simply `None` not work?

**Comment by DarkFenX** [comment]
No, quite possible it's not needed, but it's a legacy code which I just preferred to leave as-is (as usually it leads to something breaking on specific platforms in specific scenarios). This dialog was not destroyed at all before, so this is an improvement anyway.

I will experiment with setting main frame as parent and post here when I have some results.

**Comment by DarkFenX** [comment]
FYI i changed dialog's parent to be top-level window. It works alright on Linux. Will test out on Mac later today or tomorrow.

**Comment by DarkFenX** [comment]
Verified on Mac OS 10.14.6, works there as well.

**Comment by DarkFenX** [comment]
Btw supposedly this issue had caused crashes on Windows and Linux as well, just less frequently (~10% vs 100% reproducibility rate).
